readme.md
# Scrolling Note Stitcher
## What is it
The tool serves the purpose of stitching together images which are specifically from scrolling note sheets. Note that it is a Command Line Interface (CLI) tool and (mostly) has no GUI. Some applications or playback tools do not give you an opportunity to download the original notesheet or PDF and only gives you the option of a scrolling notesheet/MP4. This tool converts the notesheet video into a full length notesheet.

This tool uses your GPU for processing, as it takes quite a long time to extract many images from an MP4 file quickly, thus the processing time scales on your processing power.

## Why a custom solution?
Preferrably one would use OpenCV to stitch images, but sadly OPENCV isn't easily usable on notesheets. It often mixes "image anchors" (points of reference) since so many notes are identical. It has issues parsing the thin lines of the score and the numbering of the music staff. Even if you are building up the note sheet piece by piece and are constraining it to an area, OpenCV simlpy fails to find a similar stiching reference

## How to use
Before installing, make sure that the video has the correct name of the artist and songname. The video name is used for the final export name. After, continue with the following:

1. Add the MP4(s) to the main area in a folder named "import"
    1. The folder can be manually created if it is missing
    1. Optionally, run the program once and it will create all necessary folders for you. An error will pop-up showing that there are no video files in the import folder.
1. Launch "create_images"
    1. This converts the MP4(s) to an image set
1. Select the MP4(s) to convert
    1. Note: All files can be converted in one launch
1. Launch "create_scores" after all your MP4(s) are converted
1. Select MP4(s) to convert to notesheets
1. The notesheet will pop-up in an editor, where different properties can be adjusted
1. Adjust the properties and scroll through the sheet so that it fits properly
1. Press "S" to save or "Q" to quit
    1. The tool can use several seconds after pressing S to "show response" in the GUI, however there is activity in the CLI
    1. The tool/GUI needs to be selected when pressing S or Q

## Potential improvements
- Merge Create_images and create_scores as one launch
- Build as executable with pyfreeze
- Integrate with a python library that supports music score parsing. This way we can convert the long sheet into multiple clean sheets instead of one continious one.
- Add tool-tips for S and Q to quit or save
- Remove GUI once S or Q are pressed


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2022-07-28
### Initial release
- This is the initial release of the tool


create_images.py
# Imports
import os
import glob
from numba import jit
import cv2
import math
import shutil

# Command Line Interface (CLI) variables
BASE_DIR = os.getcwd()
IMPORT_DIR = BASE_DIR + '\\Import'
EXPORT_DIR = BASE_DIR + '\\Tmp'
SAVE_DIR = BASE_DIR + '\\Scores'
DONE_DIR = IMPORT_DIR + '\\Done'
FOLDER_LIST = [IMPORT_DIR, EXPORT_DIR, SAVE_DIR, DONE_DIR]

EXIT_CLI = 99
EXIT_PROG = 100

# Editor variables
cv_images = []
HEIGHT_MAX = 500
height_val = math.floor(HEIGHT_MAX/2)
SCALE_MAX = 100
scale_val = math.floor(SCALE_MAX/2)
DISCARD_MAX = 100
discard_val = 0
VIEWPOINT_MAX = 10
viewpoint_val = 0


def make_folder(folder):
    """ Make folders """
    if not os.path.isdir(folder):
        os.makedirs(folder)


for folder in FOLDER_LIST:
    make_folder(folder)

# Command line interface to choose files
cli = True
list_of_files = glob.glob(IMPORT_DIR+'\\*.mp4')
active_files = [0 for x in list_of_files]

assert len(active_files) > 0, "There are no files in the import folder"

while (cli):
    for i, file in enumerate(list_of_files):
        status = '(Skipping)'
        if active_files[i]:
            status = '(Parsing)'
        print(f'[{i}] {os.path.basename(file)[:-4]} {status}')
    print(f"[{len(list_of_files)}] All Files")
    print(f"[{EXIT_CLI}] Continue")
    print(f"[{EXIT_PROG}] Exit Program")
    print(f"\nType any number to activate that file for parsing,"
          f"\n{len(list_of_files)} to activate all files")
    user = input()
    parse_user = user.strip()

    try:
        int_user = int(parse_user)
        if int_user < len(list_of_files):
            active_files[int_user] = not active_files[int_user]
        elif int_user == len(list_of_files):
            active_files = [1 for x in list_of_files]
        elif int_user == EXIT_CLI:
            cli = False
        elif int_user == EXIT_PROG:
            print("Quitting")
            quit()
    except Exception as e:
        print(f'That is not a number or you have an error: {e}')

active_file_names = [file for i, file in enumerate(list_of_files) if active_files[i] == 1]

# Set up video to JPG converter
SKIP_X_FILES = 300

@jit # Jit to enable GPU processing
def extract_images(path):
    export_path = EXPORT_DIR + '\\' + os.path.basename(path)[0:-4]
    make_folder(export_path)  # Create savefolder

    capture_image = cv2.VideoCapture(path)
    frame_count = 0
    userSkipFrames = SKIP_X_FILES  # each x-th frame is captured

    # Change working dir
    os.chdir(export_path)
    i = userSkipFrames
    j = 0

    # Reading each frame
    while (True):
        con, frames = capture_image.read()
        if con:     # con will test until last frame is extracted
            if (i >= userSkipFrames):
                # Giving names to each frame and printing while extracting
                name = str(j)+'.jpg'
                print('Capturing --- '+name)
                cv2.imwrite(name, frames)
                i = 1
                j += 1
            else:
                i += 1
            frame_count = frame_count + 1
        else:
            break

    os.chdir(BASE_DIR)


# Main loop after CLI
for file in active_file_names:
    # Extract images
    extract_images(file)

    # Done with extracting, now move file
    shutil.move(file, DONE_DIR)


create_scores.py
# Imports
import os
import glob
from numba import jit
import cv2
import math

# Vars
cv_images = []

HEIGHT_MAX = 500
height_val = math.floor(HEIGHT_MAX/2)

SCALE_MAX = 100
scale_val = math.floor(SCALE_MAX/2)

DISCARD_MAX = 100
discard_val = 0

VIEWPOINT_MAX = 10
viewpoint_val = 0


def create_score(score_path):
    """
    Creates a score from a series of images.
    The images must be placed in the tmp folder.
    The images can be created using create_images.py.
    """

    # Starting OPENCV operation
    # Constants
    END_OF_HEADER = 120
    WINDOW_NAME = "Adjustment tool"

    # Create trackbar functions
    def get_height(val):
        global height_val
        if val == 0:
            val = 1
        height_val = val
        update_image()

    def get_scale(val):
        global scale_val
        if val == 0:
            val = 1
        scale_val = val
        update_image()

    def get_discard(val):
        global discard_val
        discard_val = val
        update_image()

    def get_viewpoint(val):
        global viewpoint_val
        viewpoint_val = val
        update_image()

    # Create main cv2 function
    @jit # Jit specifies GPU processing
    def update_image(ret=False):
        processed_list = []
        global cv_images

        for i, src in enumerate(cv_images):
            if i == len(cv_images)-1:
                w = src.shape[1]  # Get image width
                new_height = src.shape[0]  # Get image height
                processed_list.append(src[END_OF_HEADER:new_height, 0:w])

            elif i >= len(cv_images) - discard_val:
                pass  # Pass the last X amount of frames excluding the last one

            else:
                w = src.shape[1]  # Get image width
                new_height = END_OF_HEADER + height_val
                processed_list.append(src[END_OF_HEADER:new_height, 0:w])

        global viewpoint_val
        if viewpoint_val != 0:
            processed_list = processed_list[viewpoint_val:]

        # Stack images
        build_image = cv2.vconcat(processed_list)

        # Then scale
        cropped_h = build_image.shape[0]
        scaled_h = math.floor(cropped_h * (0.8 + 0.4*scale_val/SCALE_MAX))
        resized = cv2.resize(build_image, (w, scaled_h))

        # And show
        cv2.imshow(WINDOW_NAME, resized)

        if (ret):
            return resized

    # Start
    images = glob.glob(score_path + "\\*.jpg")
    sorted_images = sorted(images, key=lambda x: int(os.path.split(x)[-1].replace(".jpg", "")))
    print(sorted_images)
    for image in sorted_images:
        cv_images.append(cv2.imread(image))

    # Process
    cv2.namedWindow(WINDOW_NAME)  # cv2.WINDOW_NORMAL
    cv2.createTrackbar("Height", WINDOW_NAME, height_val, HEIGHT_MAX, get_height)
    cv2.createTrackbar("Scale", WINDOW_NAME, scale_val, SCALE_MAX, get_scale)
    cv2.createTrackbar("Discard", WINDOW_NAME, discard_val, DISCARD_MAX, get_discard)
    cv2.createTrackbar("Viewpoint", WINDOW_NAME, viewpoint_val, VIEWPOINT_MAX, get_viewpoint)

    update_image()

    # Wait until user presses some key
    while (cv2.waitKey() != 113 or cv2.waitKey() != 81 or
           cv2.waitKey() != 83 or cv2.waitKey() != 115):

        if (cv2.waitKey() == 83 or cv2.waitKey() == 115):
            tmp_dir = SAVE_DIR + f"\\{os.path.basename(os.path.normpath(score_path))}.jpg"
            tmp_img = update_image(True)
            cv2.imwrite(tmp_dir, tmp_img)
            cv2.destroyAllWindows()
            return True
        else:
            cv2.destroyAllWindows()
            return False


# Set up Command Line Interface (CLI)
BASE_DIR = os.getcwd()
EXPORT_DIR = BASE_DIR + "\\Tmp"
SAVE_DIR = BASE_DIR + "\\Scores"
FOLDER_LIST = [SAVE_DIR]
EXIT_CLI = 99
EXIT_PROG = 100


def make_folder(folder):
    """ Creates the input folder if it does not exist """
    if not os.path.isdir(folder):
        os.makedirs(folder)


for folder in FOLDER_LIST:
    make_folder(folder)

cli = True
list_of_folders = [x[0] for x in os.walk(EXPORT_DIR)]
list_of_folders.pop(0)  # Pop to remove parent folder
active_folders = [0 for folder in list_of_folders]

while (cli):
    for i, file in enumerate(list_of_folders):
        status = "(Skipping)"
        if active_folders[i]:
            status = "(Parsing)"
        print(f"[{i}] {os.path.basename(os.path.normpath(file))} {status}")
    print(f"[{len(list_of_folders)}] All Files")
    print(f"[{EXIT_CLI}] Continue")
    print(f"[{EXIT_PROG}] Exit Program")
    print(f"\nType any number to activate that file for parsing,"
          f"\n{len(list_of_folders)} to activate all files")
    user = input()
    parse_user = user.strip()

    try:
        int_user = int(parse_user)
        if int_user < len(list_of_folders):
            active_folders[int_user] = not active_folders[int_user]
        elif int_user == len(list_of_folders):
            active_folders = [1 for x in list_of_folders]
        elif int_user == EXIT_CLI:
            cli = False
        elif int_user == EXIT_PROG:
            print("Quitting")
            quit()
    except Exception as e:
        print(f"Not a number, or you have an error. {e}")

active_folder_names = [file for i, file in enumerate(list_of_folders) if active_folders[i] == 1]
for folder in active_folder_names:
    suc = create_score(folder)

    if suc:
        os.remove(folder)
